suites/message/puppet.go

```diff
@@ -0,0 +1,40 @@
+package message
+
+import (
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/chain-validation/drivers"
+	"github.com/filecoin-project/chain-validation/state"
+	abi_spec "github.com/filecoin-project/specs-actors/actors/abi"
+	big_spec "github.com/filecoin-project/specs-actors/actors/abi/big"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	"github.com/filecoin-project/specs-actors/actors/puppet"
+	"github.com/filecoin-project/specs-actors/actors/runtime/exitcode"
+)
+
+func MessageTest_PuppetActorInnerSends(t *testing.T, factories state.Factories) {
+	builder := drivers.NewBuilder(context.Background(), factories).
+		WithDefaultGasLimit(1_000_000).
+		WithDefaultGasPrice(big_spec.NewInt(1)).
+		WithActorState(append(drivers.DefaultBuiltinActorsState, drivers.PuppetActorState))
+
+	var actBal = abi_spec.NewTokenAmount(1_000_000_000)
+	var transferAmnt = abi_spec.NewTokenAmount(10)
+
+	t.Run("fail insufficient funds for transfer in inner send", func(t *testing.T) {
+		td := builder.Build(t)
+		defer td.Complete()
+
+		alice, _ := td.NewAccountActor(drivers.SECP, actBal)
+		bob, _ := td.NewAccountActor(drivers.SECP, big_spec.Zero())
+
+		td.ApplyFailure(td.MessageProducer.PuppetSend(drivers.PuppetActorAddress, alice, &puppet.SendParams{
+			To:     bob,
+			Value:  transferAmnt,
+			Method: builtin.MethodSend,
+			Params: nil,
+			// TODO: what should this be? Lotus returns SysErrSenderInvalid.
```

Comment:
Assuming that everything is set up correctly and `alice` is a valid sender, this should indeed be SysErrInsufficientFunds, and alice should pay gas for the failure.

`SysErrSenderInvalid` would imply the miner is at fault and be penalised, which would be the case if the top-level sender, alice, was not capable of sending the message.